funsor/montecarlo.py

```diff
@@ -2,57 +2,34 @@
 # SPDX-License-Identifier: Apache-2.0
 
 from collections import OrderedDict
-from contextlib import contextmanager
 
 from funsor.integrate import Integrate
-from funsor.interpreter import dispatched_interpretation, interpretation
-from funsor.terms import Funsor, eager
+from funsor.interpreter import StatefulInterpretation
+from funsor.terms import Funsor
 
 
-@dispatched_interpretation
-def monte_carlo(cls, *args):
+class MonteCarlo(StatefulInterpretation):
     """
     A Monte Carlo interpretation of :class:`~funsor.integrate.Integrate`
-    expressions. This falls back to :class:`~funsor.terms.eager` in other
-    cases.
-    """
-    # TODO Memoize sample statements in a context manager.
-    result = monte_carlo.dispatch(cls, *args)(*args)
-    if result is None:
-        result = eager(cls, *args)
-    return result
-
+    expressions. This falls back to the previous interpreter in other cases.
 
-# This is a globally configurable parameter to draw multiple samples.
-monte_carlo.sample_inputs = OrderedDict()
+    :param rng_key:
+    """
+    def __init__(self, *, rng_key=None, **sample_inputs):
+        self.rng_key = rng_key
```

Comment:
@fehiepsi I'm not sure whether `rng_key` should be threaded through the constructor, or whether it should instead be read-and-written in pattern handlers like `monte_carlo_integrate()` below.